### ISSUE TITLE
Consider adding Piped as a YouTube alternative

### ISSUE THREAD

**Issue by gleb09** [SimonBrazell/privacy-redirect#181]
I've just stumbled onto Piped (https://github.com/TeamPiped/Piped) (https://piped.kavin.rocks/) as a YouTube privacy-friendly frontend for YouTube. It seems to work really well compared to some Invidious instances. You should consider adding it to the "Invidious redirects" list (maybe change "Invidious redirects" to "YouTube redirects" for that). Although to note, I've noticed that it doesn't provide region locked content like Invidious does through googlevideo.

**Comment by frostrenew** [comment]
I didn't know about Piped until now but it looks promising. It's true that many Invidious instances are hit-or-miss.

**Comment by moriel5** [comment]
From what I had read, Piped relies upon NewPipe-Extractor instead of YouTube-DL, which is why it is a lot more reliable, however it hasn't fully reached feature parity yet.

In short, while Piped hasn't fully matured yet, I personally see a lot more potential in it (or a possible successor, borne out of a collaboration with the ViewTube devs).

**Comment by erazemk** [comment]
If Piped were to be added, it should probably be a separate option from Invidious redirects, since they're mutually exclusive (if you want to redirect to Piped, you probably don't want to get a random Invidious instance).

Piped also seems to be much more stable than any Invidious instance I have tested (the piped.kavin.rocks instance has worked every time I tried it).

**Comment by SimonBrazell** [comment]
> Piped also seems to be much more stable than any Invidious instance I have tested

Perhaps we could replace invidious entirely with Piped in that case? Invidious has been flakey ever since the original author stopped maintaining the official instance.

**Comment by erazemk** [comment]
> Perhaps we could replace invidious entirely with Piped in that case? Invidious has been flakey ever since the original author stopped maintaining the official instance.

Invidious probably could be removed (at least in my opinion), but I think some heavy Invidious users might oppose that.

**Comment by SimonBrazell** [comment]
> Maybe Piped could be enabled by default and Invidious disabled, but still available?

Yeah something like that, I've already got FreeTube as an advanced option / alternative, I could add a dropdown instead where you select the service you prefer, invidious, piped, freetube, etc. or something like that.

**Comment by SimonBrazell** [comment]
Should try out Piped myself before I get too carried away 🙂

**Comment by erazemk** [comment]
> Yeah something like that, I've already got FreeTube as an advanced option / alternative, I could add a dropdown instead where you select the service you prefer, invidious, piped, freetube, etc. or something like that.

That would be great

**Comment by moriel5** [comment]
@SimonBrazell Best to do that.
I have tested it, and am very impressed, however there are some unimplemented features (such as DASH).

**Comment by FireMasterK** [comment]
Hey, I'm the author of Piped, it's really nice to see so much enthusiasm! 

I'll give my thoughts here:
I think there should be a separate drop-down option like you mentioned earlier as some non features are present, such as requiring JavaScript and using cloudflare. Which I've seen people get hostile about.

Some things Piped does better than Invidious:
- Geo load balancing - I have servers in the US, Netherlands, and India currently, it will automatically chose the closest server to reduce latency.
- Better dash support - I have support for webm streams and audio too.
- Support for multi-gigabit proxying - loadbalancing isn't limited by one server's capacity.
- Switching of instances can be done in the preferences menu itself

> Should try out Piped myself before I get too carried away 🙂

Indeed!

> there are some unimplemented features (such as DASH).

This isn't true, in fact I have better dash support than Invidious!

**Comment by moriel5** [comment]
@FireMasterK Sorry about my mistake, is just that when I had tested Piped, nothing above 720p showed up, do I assumed that while technically DASH was supported (and I had no doubt about being better supported, from my experience with NewPipe), it was still unimplemented.

Perhaps I had tested Piped before DASH streaming was added?

And on another matter, I'll be adding an issue about adding a light theme to Piped (I'm not really a fan of dark themes, though I understand their importance) if it hasn't been added yet.

**Comment by SimonBrazell** [comment]
@FireMasterK I'll be sure to give it a go, and certainly consider adding it to this extension when I next have some time to spend on it.

**Comment by FireMasterK** [comment]
I have added support for /c/ and /user/ links to Piped, it should now be feasible to redirect using the same!

**Comment by PinkDev1** [comment]
Another vote for this. [Piped is taking off!](https://libredd.it/r/privacy/comments/om9h4p/piped_the_privacyfriendly_youtube/)

**Comment by Kotuklion** [comment]
[top 10 anime betrayals](https://github.com/TeamPiped/Piped-Redirects)

Just joking! Wanted to let people know that there's an alternative extension until this feature is implemented in Privacy Redirect.

**Comment by marreromarco** [comment]
Piped is more reliable than Invidious ! Could we please implement this request ?

**Comment by SimonBrazell** [comment]
I'd happily accept a PR for this feature, if someone was to do the ground work I could then somehow integrate it with the existing Invidious redirects, perhaps adding a way to switch between the two or something like that, unless someone wants to do that too... 🙂

**Comment by FireMasterK** [comment]
I actually working on this a while ago, however Piped and Invidious have a lot of different properties / url query parameters, what should be done for that?

**Comment by SimonBrazell** [comment]
@FireMasterK I'd treat Piped separately from Invidious, perhaps having a select box to switch between the two (and potentially more in future) then in the redirectYoutube method I'd check this preference and the call the appropriate method for either service.

**Comment by SimonBrazell** [comment]
@FireMasterK if you have already started with this do go ahead and submit whatever you have as a PR or direct me to a repo and I'll see if I can do some work on it tonight. 🙂

**Comment by Loki-L1130** [comment]
I think that this https://github.com/TeamPiped/Piped-Redirects/issues/11 this could also be considered while implementing.

**Comment by ghost** [comment]
As a matter of fact, I added https://piped.kavin.rocks by simply typing the URL in Invidious Instance which was displaying 'Random Instance Selected'. As intended, It redirected me from Youtube to Piped including YouTube Embeds in other sites.

This works fine both in Chrome and Firefox.

@SimonBrazell If you plan to implement Piped as a separate option from Invidious. I think you could consider about fully [redirecting Invidious Links to Piped](https://github.com/TeamPiped/Piped-Redirects/issues/11) too and also providing an [exception](https://github.com/TeamPiped/Piped-Redirects/issues/13) for allowing some Invidious Instances to be used which can be in case Piped is down or to use Invidious to Download Videos. I think that excluding URL's can be achieved by the default one provided by the Extension.

**Comment by ghost** [comment]
Any Updates on this, So Far ?

**Comment by moriel5** [comment]
@Loki-L1130 It would work, but not fully (I forget, however there are some links that fail to redirect), plus the settings won't change (the configuration code is different).

**Comment by ghost** [comment]
@moriel5 As far as I am using the setup I am using now fully forks in both Chromium and Firefox. And redirects and even replaces youtube embeds in other webpages. What do you mean by 'not fully' here ?

You mean that Invidious settings won't apply here ?

**Comment by moriel5** [comment]
@Loki-L1130 I forget what exactly does not redirect, perhaps it was channels.

Regarding settings, yeah, it does not keep them, though I believe that keeping settings for Invidious did not fully work for me either.
